### PR TITLE
Some more selection improvements to `TreeView` and `SelectingItemsControl`

### LINKED ISSUE #2660: Selection is shrunk to a single item under the cursor on right click on ListBox
But it should be several items. I've encountered this bug in AvaloniaILSpy: https://github.com/icsharpcode/AvaloniaILSpy/issues/13

See the implementation of multiselection here: https://github.com/AvaloniaUI/Avalonia/issues/206

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

A few more improvements to multiple selection on `TreeView` and `SelectingItemsControl`:

- Don't reset selection when right-clicking on an existing `TreeView` selection. Fixes #2660 
- Don't call the overload of `SetAndRaise` which takes a callback in `TreeView.SelectedItem`
  - This was only needed in `SelectingItemsControl` due to the interaction between `SelectedIndex` and `SelectedItem`, which isn't a problem here (and the need for it was removed for `SelectingItemsControl` in #2673)
- Added `SelectAll`, `UnselectAll` and `ExpandSubTree` methods to `TreeView`
- A few more improvements to `SelectingItemsControl` multiple selection:
  - Allow `SelectAll` regardless of `SelectionMode`: `SelectionMode` should only apply to user-interaction
  - Don't select multiple on shift /ctrl-right click

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #2660

### CODE REVIEW

File: tests/Avalonia.Controls.UnitTests/Primitives/SelectingItemsControlTests_Multiple.cs [hunk 1]
```diff
@@ -1026,6 +1026,46 @@ public void Right_Click_On_UnselectedItem_Should_Clear_Existing_Selection()
             Assert.Equal(1, target.SelectedItems.Count);
         }
 
+        [Fact]
+        public void Shift_Right_Click_Should_Not_Select_Mutiple()
```
Comment (MarchingCube) [inline tests/Avalonia.Controls.UnitTests/Primitives/SelectingItemsControlTests_Multiple.cs]: Typo in "mutiple" -> "multiple"

**Comment by MarchingCube** [comment]
As I mentioned on gitter, right clicking on items not always sets selection.

context-tree-view

**Comment by grokys** [comment]
@MarchingCube ok that seems to be caused by having a context menu already open. When the existing open popup detects a click outside its bounds, it [sets `e.Handled = true`](https://github.com/AvaloniaUI/Avalonia/blob/0b88151b658cc67cb8a8dd70ffff60e2a55964a2/src/Avalonia.Controls/Primitives/Popup.cs#L443). 

This was done on purpose to fix a problem according to the commit: https://github.com/AvaloniaUI/Avalonia/commit/2c9e1f8f7c0257c0fcab7984548d1f34b10203ab though unfortunately an issue isn't quoted and this was from before all changes had an associated PR :( 

Given that this is a `Popup` issue, I think it should be addressed in a separate PR.

**Review by MarchingCube — APPROVED** [review]
Apart from typos I wrote about - LGTM.

**Pull request merged**